### PR TITLE
[Qwen3.5] Add fused Triton OffsetRMSNorm

### PR DESCRIPTION
**Pull request by wwwjn**
## Summary

- Add an opt-in Triton implementation of Qwen3.5 `OffsetRMSNorm`, including fused forward, input-gradient, and weight-gradient kernels.
- Preserve the existing zero-centered parameterization, state-dict keys, FP32 accumulation, DTensor sharding contracts, CPU fallback, and `torch.compile` support.
- Add `rl_grpo_qwen3_6_27b_varlen_perf`, an 8-GPU Qwen3.6-27B RL config that uses the Qwen3.5-compatible model flavor and enables the override for both trainer and generator.

The standalone override can be enabled with:

```bash
--override.imports torchtitan.overrides.offset_rmsnorm.triton_offset_rmsnorm
```

The new RL config uses TP2 x FSDP2 for the trainer and TP4 for the generator. Full activation checkpointing and BF16 Adam states keep the 27B trainer within H100 memory.

## Why

The stock operation computes `(1 + weight) * rmsnorm(input)` as separate eager PyTorch operations. It materializes intermediates and launches separate square, reduction, reciprocal-square-root, add, multiply, and cast operations. The Triton path computes FP32 RMS statistics and the offset scaling in one forward kernel. Its custom backward similarly fuses the input-gradient work and reduces the weight gradient in Triton.

## Numerics and compatibility

- Forward, input gradient, and weight gradient are checked against both an FP32 PyTorch reference and an FP64 golden reference.
- Coverage includes FP32, BF16, FP16, non-power-of-two dimensions, Qwen hidden sizes 4096/5120, near-zero variance, and zero variance.
- `torch.library.opcheck` validates schema, FakeTensor, and autograd registration.
- A fullgraph `torch.compile` forward/backward test passes.
- The override preserves checkpoint layout and applies to every Qwen3.5 `OffsetRMSNorm` config.
- The fused inference path successfully captures and replays inside vLLM CUDA Graphs.

## Performance

All comparisons use the same code/configuration except for the override.

### Pretraining

Qwen3.5-27B, 4x H100, BF16, TP2 x FSDP2, sequence length 4096, 8192 unique tokens/step, full activation checkpointing, C4 packed text, and BF16 Adam states. CUDA Graphs are disabled because variable packed-document mask/GDN metadata currently changes structure between batches. Statistics are steps 5-34 after warmup.

| Metric | Stock | Triton | Change |
| --- | ---: | ---: | ---: |
| Throughput/GPU | 1156.69 tok/s | 1235.30 tok/s | **+6.80%** |
| End-to-end step | 1.7707 s | 1.6583 s | **-6.35%** |
| MFU | 18.63% | 19.89% | **+1.27 pp** |
| Max reserved memory | 89.93 GiB | 89.93 GiB | unchanged |

A second paired short run measured +12.22% throughput; the table reports the conservative longer run. Both 34-step C4 smoke runs followed the same loss trend and ended at 3.2830 (stock) and 3.2800 (Triton). Standard FP32 Adam states OOM while materializing optimizer state on this topology, so both sides use `fused_opt_states_bf16`.

### Inference

Qwen3.5-compatible 27B model, 4x H100, TP4, BF16, `torchtitan.experiments.rl.generate`, one sequence, 1024 generated tokens, and vLLM `FULL_AND_PIECEWISE` CUDA Graphs with decode capture size 1. The model was random-initialized to isolate execution performance from the unrelated HF-to-TP4 checkpoint-loading path.

| Metric | Stock | Triton | Change |
| --- | ---: | ---: | ---: |
| Decode throughput | 51.814 tok/s | 60.227 tok/s | **+16.24%** |

## Validation

```text
pytest tests/unit_tests/cpu/test_triton_offset_rmsnorm_override.py -q
4 passed

pytest tests/unit_tests/gpu/test_triton_offset_rmsnorm_override.py -q
5 passed, 9 subtests passed

pytest torchtitan/experiments/rl/tests/test_generator.py -q
26 passed, 1 skipped

SKIP=pyrefly-check pre-commit run --all-files
all enabled hooks passed
```

The repository-wide Pyrefly hook was skipped locally because this environment lacks the optional `torch_checkpointing` and `attn_gym` packages and uses a PyTorch build with known API skew; the changed Python files pass AST, flake8, ufmt, pydoclint, and codespell.

### CODE REVIEW

File: torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py [hunk 1]
```diff
@@ -1066,3 +1067,39 @@ def rl_grpo_qwen3_5_debug_varlen_batch_invariant() -> Controller.Config:
         config.generator, debug=_BATCH_INVARIANT_DEBUG
     )
     return config
+
+
+def rl_grpo_qwen3_6_27b_varlen() -> Controller.Config:
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py]: could be more clear if we add a "perf" or something to the config name

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py (hunk 1 above)
Comment (wwwjn) [reply torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py]: Renamed it to `rl_grpo_qwen3_6_27b_varlen_perf` to make the opt-in performance behavior explicit.
